File: app/src/main/java/com/example/parstagram/Post.java
package com.example.parstagram;

import com.parse.ParseFile;
import com.parse.ParseObject;
import com.parse.ParseClassName;
import com.parse.ParseUser;

import org.json.JSONArray;
import org.json.JSONException;
import org.parceler.Parcel;

@Parcel(analyze = {Post.class})
@ParseClassName("Post")
public class Post extends ParseObject {

    public static final String KEY_DESCRIPTION = "description";
    public static final String KEY_IMAGE = "image";
    public static final String KEY_USER = "user";
    public static final String KEY_CREATED_KEY = "createdAt";
    public static final String KEY_LIKES = "likes";
    public static final String KEY_USERS_WHO_LIKED = "usersWhoLiked";
    public static final String KEY_COMMENTS = "comments";

    public Post() {
        super();
    }

    public String getDescription() {
        return getString(KEY_DESCRIPTION);
    }

    public void setDescription(String description) {
        put(KEY_DESCRIPTION, description);
    }

    public ParseFile getImage() {
        return getParseFile(KEY_IMAGE);
    }

    public void setImage(ParseFile parseFile) {
        put(KEY_IMAGE, parseFile);
    }

    public ParseUser getUser() {
        return getParseUser(KEY_USER);
    }

    public ParseFile getProfilePic() {return getParseUser(KEY_USER).getParseFile("profilePic"); }

    public void setUser(ParseUser user) {
        put(KEY_USER, user);
    }

    public JSONArray getComments() { return getJSONArray(KEY_COMMENTS); }

    public void setComments(JSONArray array) { put(KEY_COMMENTS, array); }

    // Add a comment to the comment array and return the resulting array
    public JSONArray addComment(String comment) throws JSONException {
        JSONArray array = getComments().put(0, comment);
        setComments(array);
        return array;
    }

    // Functions related to liking
    public JSONArray getUsersWhoLiked() {return getJSONArray(KEY_USERS_WHO_LIKED); }

    public int isLiked(ParseUser user) throws JSONException {
        // Returns index of user if they liked the post, -1 if they haven't
        JSONArray array = getJSONArray(KEY_USERS_WHO_LIKED);
        for (int i = 0; i < array.length(); i++) {
            if (array.get(i).equals(user.getObjectId())) {
                return i;
            }
        }
        return -1;
    }

    public void setUsersWhoLiked(JSONArray array) {put(KEY_USERS_WHO_LIKED, array); }

    public int addUserToLikes(ParseUser user) {
        setUsersWhoLiked(getJSONArray(KEY_USERS_WHO_LIKED).put(user.getObjectId()));
        // Return index of user in likes array
        return getJSONArray(KEY_USERS_WHO_LIKED).length() - 1;
    }

    public void removeUserFromLikes(int index) {
        JSONArray array = getJSONArray(KEY_USERS_WHO_LIKED);
        array.remove(index);
        setUsersWhoLiked(array);
    }

    public int getLikes() { return getInt(KEY_LIKES); }

    public void addLike() { put(KEY_LIKES, getLikes() + 1); }

    public void subLike() { put(KEY_LIKES, getLikes() - 1); }

}

File: app/src/main/java/com/example/parstagram/Post.kt
package com.example.parstagram

import com.parse.ParseClassName
import com.parse.ParseFile
import com.parse.ParseObject
import com.parse.ParseUser
import org.json.JSONArray
import org.json.JSONException
import org.parceler.Parcel

@Parcel(analyze = [Post::class])
@ParseClassName("Post")
class Post : ParseObject() {
    var description: String?
        get() = getString(KEY_DESCRIPTION)
        set(description) {
            put(KEY_DESCRIPTION, description!!)
        }

    var image: ParseFile?
        get() = getParseFile(KEY_IMAGE)
        set(parseFile) {
            put(KEY_IMAGE, parseFile!!)
        }

    var user: ParseUser?
        get() = getParseUser(KEY_USER)
        set(user) {
            put(KEY_USER, user!!)
        }

    val profilePic: ParseFile?
        get() = getParseUser(KEY_USER)!!.getParseFile("profilePic")

    var comments: JSONArray?
        get() = getJSONArray(KEY_COMMENTS)
        set(array) {
            put(KEY_COMMENTS, array!!)
        }

    // Add a comment to the comment array and return the resulting array
    @Throws(JSONException::class)
    fun addComment(comment: String?): JSONArray {
        val array = comments!!.put(0, comment)
        comments = array
        return array
    }

    // Functions related to liking
    var usersWhoLiked: JSONArray?
        get() = getJSONArray(KEY_USERS_WHO_LIKED)
        set(array) {
            put(KEY_USERS_WHO_LIKED, array!!)
        }

    @Throws(JSONException::class)
    fun isLiked(user: ParseUser): Int {
        // Returns index of user if they liked the post, -1 if they haven't
        val array = getJSONArray(KEY_USERS_WHO_LIKED)
        for (i in 0 until array!!.length()) {
            if (array[i] == user.objectId) {
                return i
            }
        }
        return -1
    }

    fun addUserToLikes(user: ParseUser): Int {
        usersWhoLiked = getJSONArray(KEY_USERS_WHO_LIKED)!!.put(user.objectId)
        // Return index of user in likes array
        return getJSONArray(KEY_USERS_WHO_LIKED)!!.length() - 1
    }

    fun removeUserFromLikes(index: Int) {
        val array = getJSONArray(KEY_USERS_WHO_LIKED)
        array!!.remove(index)
        usersWhoLiked = array
    }

    val likes: Int
        get() = getInt(KEY_LIKES)

    fun addLike() {
        put(KEY_LIKES, likes + 1)
    }

    fun subLike() {
        put(KEY_LIKES, likes - 1)
    }

    companion object {
        const val KEY_DESCRIPTION = "description"
        const val KEY_IMAGE = "image"
        const val KEY_USER = "user"
        const val KEY_CREATED_KEY = "createdAt"
        const val KEY_LIKES = "likes"
        const val KEY_USERS_WHO_LIKED = "usersWhoLiked"
        const val KEY_COMMENTS = "comments"
    }
}